ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryRegistryImpl.java

```diff
@@ -333,8 +333,13 @@ private void registerPersistentQuery(
       unregisterQuery(oldQuery);
     }
 
-    // Initialize the query before it's exposed to other threads via the map/sets.
-    persistentQuery.initialize();
+    // If the old query was sandboxed, then the stop() won't stop the streams and will cause
```

Comment:
For my understanding, can you explain why stop() won't stop streams? Or maybe more broadly, why we were running these KStreams applications in a sandboxed environment? Just linking me to something would be sufficient, I wasn't able to piece it together myself easily.

Reply:
Sandboxed objects were implemented in KSQL to validate statements when some calls to Kafka were needed. The validation should not execute any real call (i.e. delete topics, create, etc). It's just a validation. Start looking at the `RequesValidator.validate()` class/method.

During a validation, the full execution path is tested too. For instance, in the case of persistent queries, the validation will check a query is removed from the query registry, new queries are being added to the registry, etc. To do that, the validation process creates a sandbox for the query registry to prevent real queries are indeed removed, stopped, etc. You can see the `QueryRegistryImpl.createSandbox()` method.

Before making the QueryRegistry copy, the `createSandbox()` method creates copies of each query as sandboxed persistent queries, which have dummy methods to start/stop the query. Look at `SandboxedPersistentQueryMetadataImpl`. So, when the `registerPersistentQuery` is called by a CREATE OR REPLACE statement, the old query to be stopped is not actually stopped. It just mimics the path execution, but nothing is altered in that streams. However, the initialize() is not sandboxed, so it attempts to initialize a new stream pointing to the state store from the old streams.

I don't understand too much about the sandboxed objects, though, so it took me time to figure out why new streams are started (in sandbox), but not stopped. I wanted to fix that, but I couldn't figure out an easy way to refactor the code.

Reply:
Ah okay, I've seen these validation code paths plenty of times before, but I didn't know what was going on in this detail. Thanks for the in-depth explanation!